presto-spi/src/main/java/com/facebook/presto/spi/SubfieldPathTokenizer.java

```diff
@@ -0,0 +1,282 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.spi;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_FUNCTION_ARGUMENT;
+import static java.lang.Character.isLetterOrDigit;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class SubfieldPathTokenizer
```

Comment:
Correct me if I'm wrong but I see some inspiration from AbstractIterator :)  If so I'd leave a comment explaining.  Or perhaps for the benefit of others using SPI we copy it over?

Reply:
@tdcmeehan You are correct :-) Originally, I copied AbstractIterator verbatim, but it pulls in some unnecessary dependencies and code style in that class doesn't match Presto. Hence, I folded the logic into SubfieldPathTokenizer.